interfaces/prompting/patterns/render.go

```diff
@@ -0,0 +1,372 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"bytes"
+	"errors"
+	"strings"
+)
+
+// variantState is the current variant of a render node.
+type variantState interface {
+	// Render renders the variant to the buffer if alreadyRendered equals 0 or
+	// the node is not a literal. The alreadyRendered parameter is the number
+	// of bytes of the variant which have already been rendered because they
+	// are unchanged from the previous variant.
+	//
+	// If alreadyRendered is greater than 0 and the associated renderNode is
+	// a literal, subtract from alreadyRendered the length in bytes of the
+	// literal and return the difference.
+	Render(buf *bytes.Buffer, alreadyRendered int) int
+	// NextVariant modifies the variant to the next state, if any remain.
+	// Returns the total length in bytes of the new variant, the number of
+	// bytes which remain unchanged since the previous variant, and true
+	// if more variants remain to be rendered.
+	NextVariant() (length, lengthUnchanged int, moreRemain bool)
+	// Length returns the total length of the rendered node in its current
+	// variant state, including all sub-nodes if the node is a seq or alt.
+	Length() int
+}
+
+// renderNode is a node which may be rendered in a particular variant state.
+type renderNode interface {
+	// InitialVariant returns the initial variant state for this node, along
+	// with its length.
+	InitialVariant() (variantState, int)
+	// NumVariants returns the number of variants this node can be rendered as (recursively).
+	NumVariants() int
+	// nodeEqual returns true if two nodes are recursively equal.
+	nodeEqual(renderNode) bool
+}
+
+// renderAllVariants renders subsequent variants to a reusable buffer.
+//
+// Each variant is a fully expanded path pattern, with one alternative chosen
+// for each group in the path pattern. The given observe closure is then called
+// on each variant, along with its index, and it can perform some action with
+// the variant, such as adding it to a data structure.
+func renderAllVariants(n renderNode, observe func(index int, variant string)) {
+	var buf bytes.Buffer
+	var moreRemain bool
+
+	v, length := n.InitialVariant()
+	lengthUnchanged := 0
+
+	for i := 0; ; i++ {
```

Comment:
This could just say
```suggestion
	for i := 0; moreRemain; i++ {
```

if moreRemain is initialized to true above, you can then drop the break statement

Reply:
Ahhh good point!